mitiq/cdr/data_regression.py

```diff
@@ -0,0 +1,294 @@
+            ]
+            training_obs_sim = calculate_observable(
+                training_circ_simulated_data, observable
+            )
+            Y_data[count] = training_obs_sim
+    return (X_data, Y_data)
+
+
+# TODO: again this function below is not really necessary, but is quite useful
+#  and follows the structure of the above format.
+
+
+def construct_circuit_data_floats(
+    circuit_data: List[Dict], observable: np.ndarray, noise_levels: int = 1,
+) -> np.ndarray:
+    """Returns training data now as two arrays of floats to be used in the
+    regression (raw_training_data, simulated_training_data)
+    Args:
+        circuit_data: list of dictionary of counts for circuit of interest
+                      and all noise levels.
+        observable: option to be passed to use defined observable function that
+                    defines how to calculate the value of an observable from
+                    the counts.
+        noise_levels: number of noise levels to be used in the regression.
+    """
+    circuit_data_floats = []
+    for result in circuit_data:
+        obs_raw = calculate_observable(result, observable)
+        circuit_data_floats.append(obs_raw)
+    return circuit_data_floats
+
+
+def find_optimal_parameters(
+    X_data: np.ndarray, y_data: np.ndarray, intercept: bool = True
+):
+    """Returns the optimal parameters calculated by computing a regression
+    on the training data.
+    Args:
+        X_data: array of noisy observables at various noise levels.
+        y_data: array of exact observables.
+        intercept: whether or not to include and intercept in the fitting
+                   function.
+    """
+    noise_levels = len(X_data[0][:])
+    X_data = X_data.T
+    if intercept:
+        params_0 = np.zeros((noise_levels + 1))
+    else:
+        params_0 = np.zeros((noise_levels))
+
+    # Here need to use some package that does the regression.
+    popt, pcov = curve_fit(
+        lambda x, *params: wrapper_fit_func(
+            x, noise_levels, params, intercept
+        ),
+        X_data,
+        y_data,
+        p0=params_0,
+    )
+    return popt
+
+
+def f_intercept(X_data: np.ndarray, params: List) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data)) + params[-1]
+
+
+def f(X_data: np.ndarray, params: np.ndarray) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits. No intercept.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data))
+
+
+def wrapper_fit_func(x, noise_levels, *args, intercept: bool = True):
```

Comment:
Can we deduce `noise_levels` from `x`? An can we deduce `intercept` comparing the `len(x)` and `len(args)` ?
```
def wrapper_fit_func(x, *args, intercept: bool = True): 
    if len(x) == len(args):   
        return x.dot(args)
    else:
        return x.dot(args[:-1]) + args[-1]
```

 In this case we can also remove `f` and` f_intercept `